config/config/src/lib.rs

```diff
@@ -219,45 +222,52 @@ impl<C: Config> Default for ConfigBuilder<C> {
     }
 }
 
-fn merge(first: Value, second: Value) -> (Value, Option<ConfigError>) {
-    let values = (first, second);
-    let (Value::Map(mut first), Value::Map(mut second)) = values else {
-        return (values.0, None);
-    };
-
-    let mut merged = BTreeMap::new();
-    // Get all unique keys from both maps
-    let keys = first
-        .keys()
-        .chain(second.keys())
-        .cloned()
-        .collect::<HashSet<_>>();
-
-    let mut error: Option<ConfigError> = None;
-
-    for key in keys {
-        let first = first.remove(&key);
-        let second = second.remove(&key);
-
-        let (value, new_error) = match (first, second) {
-            (Some(first), Some(second)) => merge(first, second),
-            (Some(first), None) => (first, None),
-            (None, Some(second)) => (second, None),
-            (None, None) => unreachable!(),
-        };
-
-        error = merge_error_opts(error, new_error);
-        merged.insert(key, value);
+fn merge(first: Value, second: Value) -> Value {
+    match (first, second) {
+        (Value::Map(mut first), Value::Map(mut second)) => {
+            let mut merged = BTreeMap::new();
+            // Get all unique keys from both maps
+            let keys = first
+                .keys()
+                .chain(second.keys())
+                .cloned()
+                .collect::<HashSet<_>>();
+        
+            for key in keys {
+                let first = first.remove(&key);
+                let second = second.remove(&key);
+        
+                let value = match (first, second) {
+                    (Some(first), Some(second)) => merge(first, second),
+                    (Some(first), None) => first,
+                    (None, Some(second)) => second,
+                    (None, None) => unreachable!(),
+                };
+        
+                merged.insert(key, value);
+            }
+        
+            Value::Map(merged)
+        },
+        (Value::Seq(first), Value::Seq(second)) => {
+            let mut first = first.into_iter();
+            let mut second = second.into_iter();
+            let mut merged = Vec::new();
+            while let Some(v) = first.next().or(second.next()) {
+                merged.push(v);
+            }
+            Value::Seq(merged)
```

Comment:
This is not the same, your suggestion chains the `Vec`s and the original code iterates both and merges them.

Reply:
It is the same. They produce identical assembly.

Reply:
I don't know if I'm missing something here but I tried it and the tests fail.

Reply:
Maybe the `or` is confusing. `second.next()` is always executed, not only when `first` is finished.

Reply:
> Maybe the `or` is confusing. `second.next()` is always executed, not only when `first` is finished.

So the loop is simultaneously iterating over first and second. Not one after the other.

Reply:
Oh, interesting ok so is the desired behavior one or the other? Not both?

Reply:
The desired behavior is: Iterate all element in `first` and `second` as long as there is still an element in either one of them.
**Example**
Let's say we have `first = [x, y, z]` and `second = [a, b, c, d]`.
Now the loop will call `next()` on both `first` and `second` each iteration and will take the value of `first.next()` or, if that's None, the value of `second.next()`. It stops as soon as both are `None`.
So the result will look like this: `merged = [x, y, z, d]`

Reply:
oh i see

Reply:
```rs
fn main() {
    let v1 = vec![1, 2, 3];
    let v2 = vec![7, 6, 6, 4];
    
    let len_v1 = v1.len();

    let v3: Vec<_> = v1.into_iter()
        .chain(v2.into_iter().skip(len_v1))
        .collect();

    println!("{:?}", v3);  // This should print: [1, 2, 3, 4]
}
```

Reply:
I think the suggested code would be the same when I used `or_else` instead of `or`.
```rust
while let Some(v) = first.next().or_else(|| second.next()) {
    merged.push(v);
}
```
But that's not what we want. That would just chain the the second Vec behind the first one. But we want to merge them.

Reply:
Yeah your suggestion works too. Which version is better?

Reply:
probs the rust iterator version.

Reply:
shouldn't we call "merge"  when both items exists? like:
```suggestion
        (Value::Seq(mut first), Value::Seq(second)) => {
            for (idx, v1) in second.into_iter().enumerate() {
                if let Some(dest) = first.get_mut(idx) {
                    let v2 = std::mem::replace(dest, Value::Unit);
                    *dest = merge(v1, v2);
                } else {
                    first.push(v1);
                }
            }
            Value::Seq(first)
        }
```

Reply:
https://godbolt.org/z/Y4aTWds3h they compile the exact same result.

Reply:
> shouldn't we call "merge" when both items exists? like:

Yeah, that's right

Reply:
the iterator version is faster it seems but im not sure why

Reply:
this one avoid bound checks, although always allocate.
```suggestion
         (Value::Seq(first), Value::Seq(second)) => {
            let mut merged = Vec::with_capacity(std::cmp::max(first.len(), second.len()));
            let mut first = first.into_iter();
            let mut second = second.into_iter();

            loop {
                match (first.next(), second.next()) {
                    (None, None) => break,
                    (Some(first), Some(second)) => merged.push(merge(first, second)),
                    (None, Some(second)) => merged.push(second),
                    (Some(first), None) => merged.push(first),
                }
            }

            Value::Seq(merged)
        }
```

Reply:
I think we'll take this. It's more readable than the previous one.